### ISSUE TITLE
Unexpected behavior (usability) in Register-PSRepository

### ISSUE THREAD

**Issue by HemantMahawar** [PowerShell/PowerShellGet#61]
_From @WernerMairl on November 18, 2016 8:27_

Hi
Working with Register-PSRepository i have found some behavior that was really unexpected.
Probably we are not talking about bugs but about usability....

1. Repository uniqueness
All the functions around PSRepository (Get, Register, Unregister) are using the Name Property/Argument to identify a Repo.
 
Seeing this API-Surface i'm assuming that the property Name is the one that implements the "unique" Key for Repos => expected behavior at the first look....

after playing a few hours with the api, i find out that internally NOT the Name must be unique but the property "SourceLocation" enforces uinqueness.

- That means i'm not be able to register the same package Source with different Names
- Means in different Powershell-Codebases that should run on the same server i must synchronize Repository NAMES....

To solve problems like that must select/filter my Repos by SourceLocation BUT the Get-PSRepo function doesn't implement filter for SourceLocation.... usability Issue, a property that is used as primary key should have some easy access....(yes i can do that with pure Powershell, but we are talking about API-Usability)

2. ErrorActionPreference is ignored!
If i call "Register-PSRepository" with a unique/random Name but a SourceLocation that is used in another Repo then the Function writes Error output.

_**PackageManagement\Register-PackageSource : The repository could not be registered because there exists a registered repository with Name 'XXXXX' and SourceLocation 'http://xxxxxxx.y.z.a:8080/api/odata'. To register 
another repository with Name 'XXXXX', please unregister the existing repository using the Unregister-PSRepository cmdlet**_

I found the following unexpected behavior here:

$ErrorctionPreference value is "Stop" in the calling script => assuming that a exception can be catched with try catch
Register-PSRepository writes the error record but it does not FOLLOWS the STOP behavior.. it IGNORES!

To enforce the "expected" behavior i must expplicitely use
Register-PSRepository ......-ErrorAction Stop

I'm not a expert in all the coding and behavior rules for PS but 
in 99% of my PS-Code i'm able to catch exceptions by using $ErrorActionPrefence on the global level
so i think that can be behavior issue inside Register-PSRepository

regards
Werner

_Copied from original issue: PowerShell/PowerShell#2717_

**Comment by WernerMairl** [comment]
Result of some investigation:
- Verified with v1.1.1.0 of PackageManagement AND PowershellGet!
- after reading some sourcecode i can image that the "problem" is cased by the NuGet-Provider and not by PackageManagement/PowershellGet
  (https://github.com/OneGet/NuGetProvider/blob/5bf937d437c5a648d516134af75b7dc7e49fe8e4/NugetLightProvider.cs, Line 290-293)

here a small sample that shows the behavior:

```
$ErrorActionPreference = "Stop";
try
{
  $uniqueName=[System.Guid]::NewGuid().ToString("N");
  Register-PSRepository -Name $uniqueName -SourceLocation "https://www.powershellgallery.com/api/v2/" -PackageManagementProvider "NuGet" -verbose
  Write-Host "We should never see this line because ErrorActionPreference is set to Stop and the repository with the given uniquename is NOT created!" ;
  $repo=Get-PSRepository -Name $uniqueName -ErrorAction Ignore; 
  if ($repo -eq $null)
  {
    Write-Host "Repo not created and creation NOT terminated with a exception!"
  }
}
Catch
{
  Write-Host "$($_)";
}

```

**Comment by Jaykul** [comment]
This works for my purposes, but I don't want to ever write this:

```posh
$Repo = @{
    Name               = "TeamName"
    SourceLocation     = "https://TeamName.pkgs.visualstudio.com/_packaging/PowerShell/nuget/v2"
    InstallationPolicy = "Trusted"
}

try {
    Set-PSRepository @Repo
} catch {
    Register-PSRepository @Repo
}
```

I want the *Set* or *Register* commands to have a `-Force` switch that causes it to _just do the thing_.

The whole problem with the underlying NuGet provider is just annoying, since as far as I can tell that was **created by this team** in an attempt to double-down on the package provider abstraction, but it feels like *abandonware* which has not been kept up (i.e. it still doesn't work with the modern v3 feeds, so it's **_NO LONGER ACTUALLY A NUGET PROVIDER_**).

If Microsoft doesn't have the (will|man)power to *finish* the PackageManagement system (i.e. [Update-Package PowerShell/PowerShellGetv2#6](https://github.com/OneGet/oneget/issues/6) which has been the major blocker to implementing Chocolatey **FOR FOUR YEARS**) or even make releases of added features (i.e. [Nuget v3 feeds PowerShell/PowerShellGetv2#326](https://github.com/OneGet/oneget/pull/326)) and keep up a _working_ provider _for their own_ package repositories -- you should probably just abandon the concept and make PowerShellGet a standalone module.

Nobody has time to chase bugs through 4 github projects as teams pass the blame ...

**Comment by ephos** [comment]
I agree with this.   I found this but wasn't actually sure if this issue should be in this repo, or the [PackageManagement/OneGet](https://github.com/OneGet/oneget) one.

Like others we've been having to put a ton of code in pipeline steps to handle the way these modules are designed and it's a huge pain.
